services/src/pro/permissions/mod.rs

```diff
@@ -114,6 +116,39 @@ impl From<DatasetId> for ResourceId {
     }
 }
 
+impl TryFrom<(String, String)> for ResourceId {
+    type Error = Error;
+
+    fn try_from(value: (String, String)) -> Result<Self> {
+        Ok(match value.0.as_str() {
+            "layer" => ResourceId::Layer(LayerId(value.1)),
+            "layer_collection" => ResourceId::LayerCollection(LayerCollectionId(value.1)),
+            "project" => {
+                ResourceId::Project(ProjectId(Uuid::from_str(&value.1).context(error::Uuid)?))
+            }
+            "dataset" => {
+                ResourceId::DatasetId(DatasetId(Uuid::from_str(&value.1).context(error::Uuid)?))
+            }
+            "model" => {
+                ResourceId::ModelId(MlModelId(Uuid::from_str(&value.1).context(error::Uuid)?))
+            }
+            _ => {
+                return Err(Error::InvalidResourceId {
+                    resource_type: value.0,
+                    resource_id: value.1,
+                })
+            }
+        })
+    }
+}
```

Comment:
I mean, we could create a new struct like `ResourceIdFromUrl` and implement a deserialize method, but I think we still need to match against the strings, unless you want to use like the debug output of the variants for matching.

Reply:
I just thought that if we had n enum for "layer", "layer_collection", etc. (value.0) then we would not have to check that we covered on of the cases, because the deserializer would do that. There is no `_ => …` case anymore.

We just would need to make sure that the UUID stuff works. But again, we could have something like…

enum Foo {
  Layer(string),
  Project(uuid),
}

then… no error could occur here, right?

Reply:
but how would you deserialize this enum from a string without matching?

Reply:
- You keep the match block
 - You match on an Enum instead of  strings
 - you get rid of the `_` case that only happens because there could be another string
 - serde will throw an error if some type is missing
 - you are done

Reply:
maybe

```Rust
#[derive(Deserialize)]
#[serde(tag = "resource_type", content = "resource_id")]
pub enum ResourceIdPath {
    Layer(LayerId),
    LayerCollection(LayerCollectionId),
    Project(ProjectId),
}

/// Adds a new permission.
#[utoipa::path(
    tag = "Permissions",
    get,
    path = "/permissions/resources/{resource_type}/{resource_id}",
    responses(
        (status = 200, description = "List of permission", body = Vec<PermissionListing>),        
    ),
    params(
        ("resource_type" = String, description = "Resource Type"),
        ("resource_id" = String, description = "Resource Id"),
        PermissionListOptions,
    ),
    security(
        ("session_token" = [])
    )
)]
async fn get_resource_permissions_handler<C: ProApplicationContext>(
    session: C::Session,
    app_ctx: web::Data<C>,
    resource_id: web::Path<ResourceIdPath>,
    options: web::Query<PermissionListOptions>,
) -> Result<web::Json<Vec<PermissionListing>>>
where
    <<C as ApplicationContext>::SessionContext as SessionContext>::GeoEngineDB: ProGeoEngineDb,
{
    // TODO: remove
    let (a, b) = match resource_id.into_inner() {
        ResourceIdPath::Layer(a) => ("layer".to_string(), a.to_string()),
        ResourceIdPath::LayerCollection(a) => ("layer_collection".to_string(), a.to_string()),
        ResourceIdPath::Project(a) => ("project".to_string(), a.to_string()),
    };

    // TODO: From<ResourceIdPath> for ResourceId
    let resource_id = ResourceId::try_from((a, b))?;
    let options = options.into_inner();

    let db = app_ctx.session_context(session).db();
    let permissions = db
        .list_permissions(resource_id, options.offset, options.limit)
        .await?;

    Ok(web::Json(permissions))
}
``